domain/relation/state/relation.go

```diff
@@ -481,6 +483,389 @@ WHERE  relation_id = $getRelation.relation_id
+			found = true
+		}
+	case 2: // Regular relation.
+		var otherAppID application.ID
+		if appIDs[0] == unitsAppID {
+			found = true
+			otherAppID = appIDs[1]
+		} else if appIDs[1] == unitsAppID {
+			found = true
+			otherAppID = appIDs[0]
+		}
+
+		// If the unit is a subordinate, check that it can enter scope in this
+		// relation.
+		if subordinate, err := st.isSubordinate(ctx, tx, unitsAppID); err != nil {
+			return errors.Errorf("checking if application is subordinate: %w", err)
+		} else if subordinate {
+			err := st.checkSubordinateUnitCanEnterScope(ctx, tx, relationUUID, unitUUID, otherAppID)
+			if err != nil {
+				return errors.Errorf("checking subordinate unit can enter scope %w", err)
+			}
+		}
+	}
+	if !found {
+		return relationerrors.UnitNotInRelation
+	}
+
+	return nil
+}
+
+// getApplicationOfUnit returns the ID of the application associated with the
+// unit.
+func (st *State) getApplicationOfUnit(
+	ctx context.Context,
+	tx *sqlair.TX,
+	unitUUID unit.UUID,
+) (application.ID, error) {
+	type getUnitApp struct {
+		ApplicationUUID application.ID `db:"application_uuid"`
+		UnitUUID        unit.UUID      `db:"uuid"`
+	}
+	args := getUnitApp{
+		UnitUUID: unitUUID,
+	}
+	stmt, err := st.Prepare(`
+SELECT &getUnitApp.*
+FROM   unit
+WHERE  uuid = $getUnitApp.uuid
+`, args)
+	if err != nil {
+		return "", errors.Capture(err)
+	}
+	err = tx.Query(ctx, stmt, args).Get(&args)
+	if errors.Is(err, sqlair.ErrNoRows) {
+		return "", relationerrors.UnitNotFound
+	} else if err != nil {
+		return "", errors.Capture(err)
+	}
+
+	return args.ApplicationUUID, nil
+}
+
+// getApplicationsInRelation gets all the applications that are in the given
+// relation.
+func (st *State) getApplicationsInRelation(
+	ctx context.Context,
+	tx *sqlair.TX,
+	uuid corerelation.UUID,
+) ([]application.ID, error) {
+	type applicationUUID struct {
+		UUID application.ID `db:"application_uuid"`
+	}
+	relUUID := relationUUID{
+		UUID: uuid.String(),
+	}
+	stmt, err := st.Prepare(`
+SELECT &applicationUUID.application_uuid
+FROM   relation_endpoint re 
+JOIN   application_endpoint ae ON re.endpoint_uuid = ae.uuid
+WHERE  re.relation_uuid = $relationUUID.uuid
+`, relUUID, applicationUUID{})
+	if err != nil {
+		return nil, errors.Capture(err)
+	}
+	var apps []applicationUUID
+	err = tx.Query(ctx, stmt, relUUID).GetAll(&apps)
+	if errors.Is(err, sqlair.ErrNoRows) {
+		return nil, relationerrors.RelationNotFound
+	} else if err != nil {
+		return nil, errors.Capture(err)
+	}
+
+	var ids []application.ID
+	for _, app := range apps {
+		ids = append(ids, app.UUID)
+	}
+
+	return ids, nil
+}
+
+// checkSubordinateUnitCanEnterScope checks that subordinate units can enter
+// scope. Under the following conditions, a subordinate unit cannot enter a
+// relation that is not with its principle application:
+//   - The relation has scope container.
+//   - The unit entering scope is a subordinate.
+//   - The other application is not a subordinate.
```

Comment:
todo: update comment. I'm confused, is the list AND, OR? a subordinate unit cannot enter scope if it's a subordinate?

I have a bad feeling I wrote that list.

Reply:
I've expanded the comment and added a detailed example of how the situation can occur below this comment. I think this situation is confusing enough that we will benefit from writing it out in full.